### DISCUSSION TITLE
Disable the @ short-hand

### DISCUSSION THREAD

**Discussion by Drammy** [alpinejs/alpine#3207] [1. Help] [answered]
I would like to use alpine alongside .net core Blazor WASM which uses the @ sign to markup c# code.

Is it possible to disable the short-hand @ usage in alpine so that there is no clash between two technologies?

**Comment by ekwoka** [comment, marked as answer]
If you use `Alpine.prefix()` it will disable shorthands.

**Reply by Drammy** [reply]
Great, thanks for the help.  I'm very new to alpine, would you mind showing me where in the docs this is so I can learn how to apply the change?

**Reply by ekwoka** [reply]
It doesn't look like it's in the docs 🤔

You can just call, either in your scripts before `Alpine.start()` or in a `alpine:init` listener:

```js
Alpine.prefix('x-')
```

This will "override" the prefix with the same thing, and in the process disable both `:` shorthand for `x-bind` and `@` shorthand for `x-on`

(Or at least should. I know it does when you actually change the prefix)

**Reply by Drammy** [reply]
Understood, thanks again.

**Reply by browner12** [reply]
Does this help with performance at all, or would you just use it when you have a collision with another package?

I'm thinking it might have less looking to do through the DOM?

**Reply by ekwoka** [reply]
Theoretically, sure, but that is unlikely to ever be meaningful. As Alpine is pretty efficient at these things. This wouldn't be the place to try to squeeze nanoseconds out, especially since it does effect DX.

**Reply by SimoTod** [reply]
Unfortunately, that is not entirely true. if the prefix stays the same, it won't disable the shorthand.
While alpine is looping through the attributes, it just does a string replacement transforming `@` to `x-on:` which doesn't get disabled (currently it's not possible. to disable it, just it doesn't work).

Looking at the code, I believe the original intention from Caleb was to make it work so the shorthand will also work with custom prefixes (https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives/x-on.js#L6) although this doesn't really happen because of a different bug (that line of code gets processed before the alpine:init event).

As a workaround, you can use a custom prefix but it means you need to change all your x- attributes.

**Reply by ekwoka** [reply]
That's definitely a better explanation. I just knew for sure that changing the prefix disabled the shorthands, but wasn't aware of the inner workings of that

**Reply by Drammy** [reply]
@SimoTod Are you saying Alpine.prefix() won't disable the shorthand?

**Reply by ekwoka** [reply]
He is saying that using `Alpine.prefix()` will disable the shorthand only if the provided prefix is NOT `x-`

**Reply by SimoTod** [reply]
It doesn't technically disable them. However, the shorthand `@` is always evaluated as `x-on:` (at least, until the current version) so if you set a different prefix, let's say `z-`, and you change all your directive to be z-data, z-on, etc, the final result is that `@` won't do anything in the Alpine context so it would be like it was disabled.

Bear in mind that, if someone ever fixes what I think it can be considered as a bug, it could stop working in a future release although it's really unlikely to happen.

**Reply by Drammy** [reply]
OK, thanks.
With web assembly in play I think Blazor is going to become quite popular..

Do you think there is likely to be a more resilient fix for the clash with blazor's @ usage?

**Reply by SimoTod** [reply]
It needs to be supported by alpine. 

One option would be for https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives.js#L158 to prepend the element. In that way custom mappings will run before the predefined one and you can override the behaviour of `@` so it doesn't do anything, for example,
```js
document.addEventListener('alpine:init', () => Alpine.mapAttributes((attr) => attr.value.startsWith('@') ? '' : attr));
```

Another option would be to wrap https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/directives/x-on.js#L6 in an event listener for alpine:init. The same code above will then work and it will also fix the issue of shorthand being disabled with custom prefix (but this might or might not be desirable).

There are probably other options but I think they all need changes in the Alpine core

**Reply by Drammy** [reply]
Yeah I was kind of meaning do you think its likely Alpine will be changed to allow coding around the clash..  Cheers for the other suggestions.
